src/lib.rs

```diff
@@ -427,55 +342,18 @@ where
     /// let subslice = volatile.index(1..);
     /// assert_eq!(subslice.index(0).read(), 2);
     /// ```
-    pub fn index<'a, I>(&'a self, index: I) -> Volatile<&'a I::Output, A>
+    pub fn index<I>(&self, index: I) -> Volatile<I::Output, ReadOnly>
```

Comment:
big shame that this isn't `const fn`.

Reply:
It is! Unfortunately, there's not much that we can do about that since we're using the non-const `get_unchecked_mut` function (see https://github.com/rust-lang/rust/issues/74265).

Reply:
I guess I don't entirely see why you're using [T] at all instead of just making your own types, but I guess that's up to you.

Reply:
The problem isn't really the `[T]`, but that we use the [`SliceIndex`](https://doc.rust-lang.org/stable/core/slice/trait.SliceIndex.html) trait for `index`. Unfortunately, trait impls cannot be const yet, but there is an open RFC for this at https://github.com/rust-lang/rfcs/pull/2632. The nice thing about using `SliceIndex` is that we can support all the range operators too.

For only `usize` indexing, we can add a manual const implementation. I just created such a method in https://github.com/rust-osdev/volatile/pull/22/commits/b525f383af7db00f392bd78e5b05390c252acce9.  It requires a few more feature gate, but it should work.

Reply:
That all still looks way more complicated than I was expecting. I don't get why `map_const` is being called.

When I did it in `voladdress` the block index coded ended up being:

```rust
  /// Indexes to the `i`th position of the memory block.
  ///
  /// ## Panics
  /// * If the index is out of bounds this will panic.
  #[inline]
  #[must_use]
  #[track_caller]
  pub const fn index(self, i: usize) -> VolAddress<T, R, W> {
    if i < C {
      unsafe { self.base.add(i) }
    } else {
      // Note(Lokathor): We force a const panic by indexing out of bounds.
      #[allow(unconditional_panic)]
      unsafe {
        VolAddress::new([usize::MAX][1])
      }
    }
  }
```
And then the same for a Series as well.

I also have this (as non-const because no const traits yet)

```rust
pub fn iter_range<RB: core::ops::RangeBounds<usize>>(
    self, r: RB,
  ) -> VolBlockIter<T, R, W> {
```

but basically I guess I don't currently support sub-slicing or splitting or things like that. The Block and Series types aren't dynamically sized at all, just const generic. If I did want to add dynamically sized regions, my impression is that just having a separate type for it, something like `VolSlice<T: Copy, R, W> { base: VolAddress<_>, length: usize }`, and then you can kinda forget about SliceIndex, which is a little poor to begin with because all the methods are nightly.